lib/spack/spack/compiler.py

```diff
@@ -223,52 +228,59 @@ def fc_version(cls, fc):
         return cls.default_version(fc)
 
     @classmethod
-    def _find_matches_in_path(cls, compiler_names, detect_version, *path):
-        """Finds compilers in the paths supplied.
-
-           Looks for all combinations of ``compiler_names`` with the
-           ``prefixes`` and ``suffixes`` defined for this compiler
-           class.  If any compilers match the compiler_names,
-           prefixes, or suffixes, uses ``detect_version`` to figure
-           out what version the compiler is.
-
-           This returns a dict with compilers grouped by (prefix,
-           suffix, version) tuples.  This can be further organized by
-           find().
+    def search_compiler_commands(cls, operating_system, *search_paths):
+        """Returns a list of commands that, when invoked, search for compilers
+        in the paths supplied.
+
+        Args:
+            operating_system (OperatingSystem): the OS requesting the search
+            *search_paths (list of paths): paths where to look for a
+                compiler
+
+        Returns:
+            (tags, commands): ``tags`` is a list of compiler tags, containing
```

Comment:
That was just because of defensive programming. The data structure used internally before was a list and order was said to be important in a couple of comments:

https://github.com/spack/spack/blob/9f799957180ddf9e710fb3386a23ee824f56b9aa/lib/spack/spack/compiler.py#L267-L270

As you ask for it, I'll look into simplifying the code using a `dict`.